**Assignment-16**

Recommendation System

In [5]:
import numpy as np
import pandas as pd

In [6]:
df = pd.read_csv('/content/anime.csv')

In [7]:
df.head()

,anime_id,name,genre,type,episodes,rating,members
0,32281,Kimi no Na wa.,"Drama, Romance, School, Supernatural",Movie,1,9.37,200630
1,5114,Fullmetal Alchemist: Brotherhood,"Action, Adventure, Drama, Fantasy, Magic, Mili...",TV,64,9.26,793665
2,28977,Gintama°,"Action, Comedy, Historical, Parody, Samurai, S...",TV,51,9.25,114262
3,9253,Steins;Gate,"Sci-Fi, Thriller",TV,24,9.17,673572
4,9969,Gintama&#039;,"Action, Comedy, Historical, Parody, Samurai, S...",TV,51,9.16,151266


In [8]:
df.shape

(12294, 7)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12294 non-null  int64  
 1   name      12294 non-null  object 
 2   genre     12232 non-null  object 
 3   type      12269 non-null  object 
 4   episodes  12294 non-null  object 
 5   rating    12064 non-null  float64
 6   members   12294 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 672.5+ KB


In [10]:
df.describe()

,anime_id,rating,members
count,12294.000000,12064.000000,1.229400e+04
mean,14058.221653,6.473902,1.807134e+04
std,11455.294701,1.026746,5.482068e+04
min,1.000000,1.670000,5.000000e+00
25%,3484.250000,5.880000,2.250000e+02
50%,10260.500000,6.570000,1.550000e+03
75%,24794.500000,7.180000,9.437000e+03
max,34527.000000,10.000000,1.013917e+06


In [11]:
df.isnull().sum()

,0
anime_id,0
name,0
genre,62
type,25
episodes,0
rating,230
members,0


In [12]:
df.duplicated().sum()

np.int64(0)

In [13]:
df.dtypes

,0
anime_id,int64
name,object
genre,object
type,object
episodes,object
rating,float64
members,int64


In [14]:
df.columns

Index(['anime_id', 'name', 'genre', 'type', 'episodes', 'rating', 'members'], dtype='object')

In [15]:
features = ['genre','rating']

In [16]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
le.fit_transform(df['genre'])
df['genre'] = le.fit_transform(df['genre'])

In [17]:
features

['genre', 'rating']

In [18]:
df[features].head()

,genre,rating
0,2686,9.37
1,161,9.26
2,534,9.25
3,3240,9.17
4,534,9.16


In [19]:
from sklearn.preprocessing import StandardScaler

In [20]:
scaler = StandardScaler()

In [21]:
X_scaled = scaler.fit_transform(df[features])

In [22]:
X_scaled[:5]

array([[ 0.78573595,  2.82077308],
       [-1.95016038,  2.7136341 ],
       [-1.54600619,  2.70389419],
       [ 1.38600786,  2.62597493],
       [-1.54600619,  2.61623502]])

In [23]:
X_scaled

array([[ 0.78573595,  2.82077308],
       [-1.95016038,  2.7136341 ],
       [-1.54600619,  2.70389419],
       ...,
       [ 1.0208605 , -1.55244557],
       [ 1.0208605 , -1.45504649],
       [ 1.0208605 , -0.98753091]])

In [28]:
import numpy as np

np.isnan(X_scaled).sum()

np.int64(230)

In [29]:
df[features].isnull().sum()

,0
genre,0
rating,230


In [30]:
df.duplicated().sum()

np.int64(0)

In [35]:
df['rating'].fillna(df['rating'].mean())

,rating
0,9.37
1,9.26
2,9.25
3,9.17
4,9.16
...,...
12289,4.15
12290,4.28
12291,4.88
12292,4.98


In [36]:
df.rating

,rating
0,9.37
1,9.26
2,9.25
3,9.17
4,9.16
...,...
12289,4.15
12290,4.28
12291,4.88
12292,4.98


In [37]:
df[features].isnull().sum()

,0
genre,0
rating,0


In [38]:
X_scaled

array([[ 0.78573595,  2.82077308],
       [-1.95016038,  2.7136341 ],
       [-1.54600619,  2.70389419],
       ...,
       [ 1.0208605 , -1.55244557],
       [ 1.0208605 , -1.45504649],
       [ 1.0208605 , -0.98753091]])

In [42]:
df.duplicated().sum()

np.int64(0)

In [39]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(df[features])

In [40]:
from sklearn.metrics.pairwise import cosine_similarity

cosine = cosine_similarity(X_scaled)

In [43]:
cosine_similarity = pd.DataFrame(cosine, index=df['anime_id'], columns=df['anime_id'])

In [44]:
from sklearn.metrics.pairwise import cosine_similarity

cosine = cosine_similarity(X_scaled)

In [45]:
cosine

array([[ 1.        ,  0.63104039,  0.70768492, ..., -0.66253677,
        -0.63976631, -0.48318443],
       [ 0.63104039,  1.        ,  0.99466698, ..., -0.99914699,
        -0.99993615, -0.98409242],
       [ 0.70768492,  0.99466698,  1.        , ..., -0.99807765,
        -0.99576897, -0.96052088],
       ...,
       [-0.66253677, -0.99914699, -0.99807765, ...,  1.        ,
         0.99954985,  0.97591661],
       [-0.63976631, -0.99993615, -0.99576897, ...,  0.99954985,
         1.        ,  0.982022  ],
       [-0.48318443, -0.98409242, -0.96052088, ...,  0.97591661,
         0.982022  ,  1.        ]])

In [58]:
def recommend_anime(anime_id):
    index = df[df['anime_id'] == anime_id].index[0]

    distances = sorted(
        list(enumerate(cosine[index])),
        reverse=True,
        key=lambda x: x[1]
    )

    for i in distances[1:6]:
        print(df.iloc[i[0]]['name'])
        print(df.iloc[i[0]]['rating'])
        print(df.iloc[i[0]]['genre'])
        print(df.iloc[i[0]]['type'])

In [59]:
recommend_anime(1)

Doraemon Movie 20: Nobita no Uchuu Hyouryuuki
7.29
1291
Movie
Doraemon Movie 01: Nobita no Kyouryuu
7.4
1199
Movie
Daisetsusan no Yuusha Kibaou
7.13
1421
Special
Hyakko OVA
6.82
1676
OVA
Mobile Suit Gundam Thunderbolt: December Sky
7.92
776
Movie


In [67]:
def recommend_anime(anime_id, threshold):
    index = df[df['anime_id'] == anime_id].index[0]

    distances = sorted(
        list(enumerate(cosine[index])),
        reverse=True,
        key=lambda x: x[1]
    )
    recommended_anime = []
    for i in distances[1:]:
       if i[1] >= threshold:
        recommended_anime.append(df.iloc[i[0]]['name'])

    return recommended_anime

In [68]:
recommend_anime(1, 0.8)

['Doraemon Movie 20: Nobita no Uchuu Hyouryuuki',
 'Doraemon Movie 01: Nobita no Kyouryuu',
 'Daisetsusan no Yuusha Kibaou',
 'Hyakko OVA',
 'Mobile Suit Gundam Thunderbolt: December Sky',
 'Eyeshield 21',
 'GR: Giant Robo',
 'Trigun: Badlands Rumble',
 'Mutsu Enmei Ryuu Gaiden: Shura no Toki',
 'Wellber no Monogatari: Sisters of Wellber Zwei',
 'JoJo no Kimyou na Bouken: Stardust Crusaders 2nd Season',
 'D.Gray-man Hallow',
 'Hunter x Hunter OVA',
 'Hunter x Hunter: Greed Island Final',
 'Hellsing',
 'Gundam Build Fighters Try',
 'Dae Jang Geum: Jang Geum&#039;s Dream',
 'Tiger &amp; Bunny',
 'Owari no Seraph: Nagoya Kessen-hen',
 'Gintama: Jump Festa 2014 Special',
 'Chiisana Ahiru no Ooki na Ai no Monogatari: Ahiru no Kwak',
 'Mobile Suit Gundam Seed',
 'Persona 3 the Movie 3: Falling Down',
 'Negibouzu no Asatarou',
 'City Hunter 2',
 'Ramayana: The Legend of Prince Rama',
 'Tales of Phantasia The Animation',
 'Prince of Tennis: The National Tournament',
 'Code Geass: Boukoku no Ak

In [69]:
recommend_anime(1, 0.9)

['Doraemon Movie 20: Nobita no Uchuu Hyouryuuki',
 'Doraemon Movie 01: Nobita no Kyouryuu',
 'Daisetsusan no Yuusha Kibaou',
 'Hyakko OVA',
 'Mobile Suit Gundam Thunderbolt: December Sky',
 'Eyeshield 21',
 'GR: Giant Robo',
 'Trigun: Badlands Rumble',
 'Mutsu Enmei Ryuu Gaiden: Shura no Toki',
 'Wellber no Monogatari: Sisters of Wellber Zwei',
 'JoJo no Kimyou na Bouken: Stardust Crusaders 2nd Season',
 'D.Gray-man Hallow',
 'Hunter x Hunter OVA',
 'Hunter x Hunter: Greed Island Final',
 'Hellsing',
 'Gundam Build Fighters Try',
 'Dae Jang Geum: Jang Geum&#039;s Dream',
 'Tiger &amp; Bunny',
 'Owari no Seraph: Nagoya Kessen-hen',
 'Gintama: Jump Festa 2014 Special',
 'Chiisana Ahiru no Ooki na Ai no Monogatari: Ahiru no Kwak',
 'Mobile Suit Gundam Seed',
 'Persona 3 the Movie 3: Falling Down',
 'Negibouzu no Asatarou',
 'City Hunter 2',
 'Ramayana: The Legend of Prince Rama',
 'Tales of Phantasia The Animation',
 'Prince of Tennis: The National Tournament',
 'Code Geass: Boukoku no Ak

In [70]:
recommend_anime(1, 0.95)

['Doraemon Movie 20: Nobita no Uchuu Hyouryuuki',
 'Doraemon Movie 01: Nobita no Kyouryuu',
 'Daisetsusan no Yuusha Kibaou',
 'Hyakko OVA',
 'Mobile Suit Gundam Thunderbolt: December Sky',
 'Eyeshield 21',
 'GR: Giant Robo',
 'Trigun: Badlands Rumble',
 'Mutsu Enmei Ryuu Gaiden: Shura no Toki',
 'Wellber no Monogatari: Sisters of Wellber Zwei',
 'JoJo no Kimyou na Bouken: Stardust Crusaders 2nd Season',
 'D.Gray-man Hallow',
 'Hunter x Hunter OVA',
 'Hunter x Hunter: Greed Island Final',
 'Hellsing',
 'Gundam Build Fighters Try',
 'Dae Jang Geum: Jang Geum&#039;s Dream',
 'Tiger &amp; Bunny',
 'Owari no Seraph: Nagoya Kessen-hen',
 'Gintama: Jump Festa 2014 Special',
 'Chiisana Ahiru no Ooki na Ai no Monogatari: Ahiru no Kwak',
 'Mobile Suit Gundam Seed',
 'Persona 3 the Movie 3: Falling Down',
 'Negibouzu no Asatarou',
 'City Hunter 2',
 'Ramayana: The Legend of Prince Rama',
 'Tales of Phantasia The Animation',
 'Prince of Tennis: The National Tournament',
 'Code Geass: Boukoku no Ak

In [71]:
r1 = recommend_anime(1, 0.8)
r2 = recommend_anime(1, 0.9)
r3 = recommend_anime(1, 0.95)

In [72]:
print(len(r1))
print(len(r2))
print(len(r3))

2905
2111
1521


In [73]:
print("Threshold 0.8:", len(r1), "recommendations")
print("Threshold 0.9:", len(r2), "recommendations")
print("Threshold 0.95:", len(r3), "recommendations")

Threshold 0.8: 2905 recommendations
Threshold 0.9: 2111 recommendations
Threshold 0.95: 1521 recommendations


Performance Analysis:

When the similarity threshold increases, the number of recommendations decreases. At 0.80, the system gives 2905 recommendations, while at 0.90 it gives 2111 and at 0.95 it gives 1521. A higher threshold gives more similar recommendations but fewer results. A lower threshold gives more results but may include less similar anime.

Areas for Improvement:

The recommendation system can be improved by using more features such as anime type, number of episodes, and user ratings. Better encoding methods and more suitable similarity thresholds could also improve recommendation quality.

User-based: Finds users with similar preferences and recommends items they liked.
Item-based: Finds items that are similar to items the user already liked.

Collaborative Filtering is a recommendation technique that uses user preferences, ratings, or behavior to recommend items. It finds similar users or items and recommends items based on those similarities.